### DISCUSSION TITLE
[indexer] Get ready to nearcore 1.26.0 and NEAR Indexer Framework 1.26.0

### DISCUSSION THREAD

**Discussion by khorolets** [near/nearcore#6656] [Node Public Interfaces]
# TL;DR `5d824f6b1c98a05cd51ba01c464f5aad7ec4aac8`
---

- `testnet` Apr 20, 2022
- `mainnet` May 18, 2022

*This release will make a protocol upgrade, your nodes might get stuck if you don't upgrade them before the protocol upgrade happens.*

---

## NEAR Lake Framework

We want to remind you that we've [released NEAR Lake Framework](https://gov.near.org/t/announcement-near-lake-framework-brand-new-word-in-indexer-building-approach/17668). If you don't have a hard requirement to use NEAR Indexer Framework and run `nearcore` node for your indexer, we encourage you to migrate to [NEAR Lake Framework](https://github.com/near/near-lake-framework).

NEAR Lake Framework allows you to build a lightweight indexer which doesn't require a lot of resources. This will:
- decrease your spending on infrastructure
- don't require updates on every `nearcore` release

Please, reach out to us if you have questions about migration, we will be happy to help.

---

## Commit hash

[`c7eaf26ba3e934a1ee76af2d2ac4266fa276d28c`](https://github.com/near/nearcore/commit/c7eaf26ba3e934a1ee76af2d2ac4266fa276d28c)

## Expected issues

NEAR Indexer Framework contains a breaking change that might require updates to the code of your indexers.

1. The field `state_changes` is moved from the root of `StreamerMessage` to the `IndexerShard.state_changes` and now contains only changes related to the specific shard.

  ```
  BEFORE
  
  streamer_message.state_changes
  
  AFTER
  
  streamer_message.shards[i].state_changes
  ```
  
  **The changes in your indexer are required if you use `state_changes` in your logic. Otherwise, you don't need to do anything**

2. All the indexer-related structures like `StreamerMessage` are extracted from the `near-indexer` crate into a separate crate [`near-indexer-primitives`](https://github.com/near/nearcore/tree/master/chain/indexer-primitives) (and it should be published to crates.io soon enough). 
   `near-indexer` crate exposes the `near_indexer_primitives`, so you don't have to import the primitives crate separately.

3. Build might fail with errors like:

```
    |
115 |         let sync_status_log = Some(display_sync_status(sync_status, head, genesis_height));
    |             --------------- move occurs because `sync_status_log` has type `Option<std::string::String>`, which does not implement the `Copy` trait
...
160 |             paint(ansi_term::Colour::Yellow, sync_status_log),
    |                                              ^^^^^^^^^^^^^^^
    |                                              |
    |                                              value moved here
    |                                              value used here after move
```

This is caused by the bug in the older `tracing` crate.

The fix is to run 

```
$ cargo update -p tracing
```

## Reference

See how we did the upgrade on the [NEAR Indexer for Explorer side](https://github.com/near/near-indexer-for-explorer/pull/272/files)

---

Feel free to leave your questions in the comments below.

**Comment by tifrel** [comment]
Undocumented (breaking) changes:

- Bumping some dependency versions ([compare here](https://github.com/near/near-indexer-for-explorer/pull/272/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542))
- Bumping `rustc` to `1.60.0`

**Comment by khorolets** [comment]
FYI I've added the fix to another known issue when your indexer build fails.
